src/System.Private.Uri/src/System/DomainNameHelper.cs

```diff
@@ -505,33 +509,41 @@ internal static unsafe string UnicodeEquivalent(char* hostname, int start, int e
         //  DNS specification [RFC 1035]. We use our own variant of IsLetterOrDigit
         //  because the base version returns false positives for non-ANSI characters
         //
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
         private static bool IsASCIILetterOrDigit(char character, ref bool notCanonical)
         {
-            if ((character >= 'a' && character <= 'z') || (character >= '0' && character <= '9'))
+            if ((uint)(character - 'a') <= 'z' - 'a' || (uint)(character - '0') <= '9' - '0')
```

Comment:
Out of curiosity, why is it that this is faster than the original implementation. Are comparisons particularly costly?

Reply:
The code ends up being a bit larger, e.g.
```C#
public class C
{
    public bool M1(char c) => (uint)(c - '0') <= '9' - '0';
    public bool M2(char c) => c >= '0' && c <= '9';
}
```
results in:
```
C.M1(Char)
    L0000: and edx, 0xffff
    L0006: add edx, 0xffffffd0
    L0009: cmp edx, 0x9
    L000c: setbe al
    L000f: movzx eax, al
    L0012: ret

C.M2(Char)
    L0000: movzx eax, dx
    L0003: cmp eax, 0x30
    L0006: jl L0012
    L0008: cmp eax, 0x39
    L000b: setle al
    L000e: movzx eax, al
    L0011: ret
    L0012: xor eax, eax
    L0014: ret
```
https://sharplab.io/#v2:EYLgtghgzgLgpgJwDQxNMAfAAgZgARYBMeAwgLABQA3pXnQfsAPZMA2eAsgIwAUAxgAsICPHwCUeALwA+PDwCuASwB2MMfzwBaPAHIADDokAeSboCcOrboMBuWvVx5mbToX5CR4qbL55pp/UsAMiDRPBNzHTsKAF8gA=

But the bigger deal is the conditional jump.  The first one doesn't have any.  The second does, and any time you have conditionals, you run the risk of branch predictors getting it wrong, which then incurs non-trivial cost.

That said, this was only a minor improvement.  The main improvement came from the inlining, followed by removal of the extra comparisons in that first loop.

Reply:
Very interesting, thanks.